src/coordinate_systems.jl

```diff
@@ -0,0 +1,17 @@
+
+"""
+    CoordinateSystem
+
+A trait for representing coordinate systems.
+"""
+abstract type CoordinateSystem end
+
+struct ImageCoordinates <: CoordinateSystem end
+
+@inline function CoordinateSystem(::Type{T}) where {T}
+    if parent_type(T) <: T
+        return ImageCoordinates()
+    else
+        return CoordinateSystem(parent_type(T))
+    end
+end
```

Comment:
I'm not sure if this approach to inheriting the `CoordinateSystem` trait is the best because this approach seems like there will need to be a unique array type for each coordinate system.
We could dispatch on dimension name combinations, (`CoordinateSystem(::Tuple{StaticSymbol{:left},StaticSymbol{:bottom}}) = ImageCoordinates()`, but that would require defining a lot of dispatch combinations for dimension names.
We could also search for it in the `MetaArray` so `attach_metdata(::AbstractArray, ::CordinateSystem)` would be all that's necessary to assign a non-default coordinate system.
I'm assuming we would want the default coordinate system to be `ImageCoordinates` since one common endpoint of most spatial data would be an image plot.
We could also just have `UnkownCoordinateSystem`

Reply:
It would also be problematic for slice-views, wouldn't it?

What about piling on to `spacedirections`? That was my intended target for coordinate systems. In addition to specifying coordinate vectors one could use named axes to indicate the coordinates.

Reply:
>What about piling on to spacedirections? That was my intended target for coordinate systems. In addition to specifying coordinate vectors one could use named axes to indicate the coordinates.

Do you mean something like [this code](https://github.com/JuliaNeuroscience/AnatomicalTraits.jl/blob/1e901ea5a158b7befc3abfbca60c21d3ae3704b4/src/orientation.jl#L48) where we get a name indicated orientation along each dimension?

Reply:
Something like that. I always find it hard to write traits without pairing it with implementation of real-world code.

Reply:
I've been using that code to take the coordinate vectors along each axis from NIfTI files to derive the most closely aligned euclidean axis names (e.g., :P, :R, :S). I've seen `orientation` used in the geometry libraries to derive similar names from types. So one application for a coordinate system trait would be translating what comes out of `_dir2ori(spacedirections(x)) -> Tuple{Int,Int,Int}` given a coordinate system trait. Perhaps it's that's too specific of an application.

I've also been trying to figure out how to codify a few standard brain coordinate systems succinctly, where things like the coordinate system's origin differ and is entirely independent of what segment of the brain is actively being viewed.

I know we could just throw this all in a metadata dictionary but features of coordinate system are typically static (until warped into a new one with a new array or view).

Reply:
Sorry, can you paste an example of the output? It's a little hard to tell what it's supposed to return.

Here's a ficticious one for `spacedirections` on an array that happens to be sampled skewed relative to real-world coordinates:
```julia
julia> spacedirections(A)
((0.4 μm, 0.3 μm, 0.0 μm), (0.0 μm, 0.5 μm, 0.0 μm), (0.0 μm, 0.0 μm, 1.4 μm))
```
Many [OPM-based microscopes](https://www.osapublishing.org/oe/fulltext.cfm?uri=oe-16-25-20306&id=175008) (of which there are now many variants) return skewed arrays, where the array dimensions are not orthogonal to one another. Examples like that are the reason that `spacedirections` has always been documented to return a tuple-of-tuples, so that any affine-mappable transformation can be represented. It's inspired by the [NRRD format](http://teem.sourceforge.net/nrrd/format.html#spacedirections) which has always been a good model for "simple" formats that do a good job of encoding geometry. (DICOM is not simple...)

Reply:
As I read your post again, it sounds like NIfTI also supplies coordinate vectors, and in fact it seems quite close to the design of `spacedirections`. What's missing, then? Are you looking for a simpler interpretation than full vectors? As the skewed example shows, you can't do that and remain general enough to handle real-world imaging devices.

Reply:
The output of `orientation` here is `Tuple{Symbol,Symbol,Symbol}`.
For NIfTI, sagittal, coronal, and axial each correspond to one of the three positions in the axis vector.
We derive "left" or "right" based on if it's positive or negative.
So a perfectly aligned image in radiological orientation (:L,:A,:S) would have a `spacedirection` return `((1, 0, 0), (0, -1, 0), (0, 0, -1))`.
So a perfectly aligned image in neurologic orientation (:R,:A,:S) would have a `spacedirection` return `((-1, 0, 0), (0, -1, 0), (0, 0, -1))`.
The (messy) code I linked to iterates through all of the possible orientations and returns the one that aligns best with the images space directions.
I figured that sort of operation would be necessary for a variety of imaging devices and the only thing that would need to change is what names the space directions map to.
Since I've seen `orientation(x) -> Tuple{Symbol,Symbol,Symbol}` used elsewhere I thought that was a safe assumption, but I could be going down a rabbit hole here.

Reply:
Don't get me wrong, it is nice to have a named, easy-to-think-about coordinate system when applicable. If that's the goal I can support it, I am now just getting clear on the the difference between this and `spacedirections`. We probably need both, since one is not a substitute for the other.

Reply:
The docstring could be a bit more expansive, however. `spacedirections` is pretty clear about what it is, but *anything* fits under CoordinateSystem according to its docstring. The one certainty is it needs to get vastly more specific about what it actually encodes.